tools/lsp/server_loop.rs

```diff
@@ -352,20 +385,56 @@ pub async fn reload_document(
     Ok(())
 }
 
-/// return the token, and the offset within the file
-fn token_descr(
+fn get_document_and_offset(
     document_cache: &mut DocumentCache,
     text_document: lsp_types::TextDocumentIdentifier,
     pos: Position,
-) -> Option<(SyntaxToken, u32)> {
+) -> Option<(&i_slint_compiler::object_tree::Document, u32)> {
     let o = document_cache.newline_offsets.get(&text_document.uri)?.get(pos.line as usize)?
         + pos.character as u32;
 
     let doc = document_cache.documents.get_document(&text_document.uri.to_file_path().ok()?)?;
-    let node = doc.node.as_ref()?;
-    if !node.text_range().contains(o.into()) {
-        return None;
+    doc.node.as_ref()?.text_range().contains(o.into()).then(|| (doc, o))
+}
+
+fn element_contains(element: &i_slint_compiler::object_tree::ElementRc, offset: u32) -> bool {
+    element.borrow().node.as_ref().map_or(false, |n| n.text_range().contains(offset.into()))
+}
+
+fn element_at_position(
+    document_cache: &mut DocumentCache,
+    text_document: lsp_types::TextDocumentIdentifier,
+    pos: Position,
+) -> Result<i_slint_compiler::object_tree::ElementRc, String> {
+    let (doc, offset) = get_document_and_offset(document_cache, text_document, pos)
+        .ok_or("Document not found.".to_string())?;
```

Comment:
```suggestion
        .ok_or_else(|| "Document not found.".to_string())?;
```

So that there are no allocations in the non-error case.  I think clippy warns about that, doesn't it?
(there are other occurences)

Reply:
Also fixed this.